### ISSUE TITLE
sql: DROP does not properly mark tables as deleted

### ISSUE THREAD

**Issue by kaeverens** [cockroachdb/cockroach#9318]
- What did you do?

dropped a table.

- What did you expect to see?

root@178.62.29.223:26257> show tables;
```
+-------+
| Table |
+-------+
| logs  |
+-------+
(1 row)
root@178.62.29.223:26257> drop table logs;
DROPPED
root@178.62.29.223:26257> show tables;
+-------+
| Table |
+-------+
+-------+
(0 rows)
```

- What did you see instead?

```
root@178.62.29.223:26257> show tables;
+-------+
| Table |
+-------+
| logs  |
+-------+
(1 row)
root@178.62.29.223:26257> drop table logs;
pq: table "logs" has been deleted
root@178.62.29.223:26257> show tables;
+-------+
| Table |
+-------+
| logs  |
+-------+
(1 row)
```

about the node:
I160913 13:50:53.910615 1 util/log/clog.go:1019  [config] file created at: 2016/09/13 13:50:53
I160913 13:50:53.910615 1 util/log/clog.go:1019  [config] running on machine: cdb1
I160913 13:50:53.910615 1 util/log/clog.go:1019  [config] binary: CockroachDB beta-20160908 (linux amd64, built 2016/09/08 15:33:58, go1.7)
I160913 13:50:53.910615 1 util/log/clog.go:1019  [config] arguments: [cockroach start --host=178.62.29.223 --http-addr=127.0.0.1 --ca-cert=ca.cert --cert=node1.cert --key=node1.key]
I160913 13:50:53.910615 1 util/log/clog.go:1019  line format: [IWEF]yymmdd hh:mm:ss.uuuuuu goid file:line msg
I160913 13:50:53.910607 1 cli/start.go:342  CockroachDB beta-20160908 (linux amd64, built 2016/09/08 15:33:58, go1.7)

**Comment by knz** [comment]
cc @andreimatei @vivekmenezes

**Comment by vivekmenezes** [comment]
https://github.com/cockroachdb/cockroach/blob/develop/sql/testdata/drop_table

has a test validating that this functionality works. We clearly have another problem here. Are you able to reproduce this reliably? What was the size of the table in terms of the number of rows

`SELECT COUNT(*) FROM logs`

Did the table eventually get dropped (delayed drop)?

thanks

**Comment by knz** [comment]
@kaeverens Thanks for reporting this btw.

Could you also remind us which version of CockroachDB you are using?

**Comment by kaeverens** [comment]
I'm sorry - I got frustrated with it after a while and destroyed the database. But, this was after an out-of-memory crash, so I think the database thought the drop table was still in the transaction the database had open at the time of the crash.

the version is beta-20160908

**Comment by knz** [comment]
Ok thanks for clarifying. Please tell us more about the sequence of events.
My current understanding is as follows:
- you tried to insert data
- the node failed with an OOM error
- subsequently you issued a DROP statement
- then you observed the table was not really dropped.

Is this sequence accurate? Or did you rather issue the DROP statement before the out-of-memory error?

**Comment by kaeverens** [comment]
you have it right.

**Comment by andreimatei** [comment]
@kaeverens did you perform the `DROP` as part of a transaction?

**Comment by kaeverens** [comment]
no. I have not used transactions at all in any of my tests. This was all straightforward one-off commands.

**Comment by andreimatei** [comment]
I think what's going on is expected. Your crashed node had a lease on a schema version of that table. The lease was never released, so it'll stick around for ~5m and during this time it'll prevent the schema change that does the dropping from running.

I suspect you did a `DROP` that you didn't paste above. I'm curious what the result of this `DROP` was... I think it should've blocked for 5min, but you somehow abandoned that session. I think then you did another `DROP` which errored with `pq: table "logs" has been deleted`. It's debatable what this error should be exactly, but I think it's fine. If you would've tried an DML statement, I think you would've gotten a "table doesn't exist" error which you probably would've liked more.

`SHOW TABLES` unfortunately has no understanding of tables that are pending deletion. We should probably improve this by showing an asterisk or a note or smth. But otherwise the result is technically correct - the table does exist and there are even transactions in the system that can still use it. It should go away in 5min though. I'll rename the issue to suggest this `SHOW TABLES` improvement if you agree.

**Comment by a-robinson** [comment]
Does nothing ever garbage collect tables that are pending deletion after a crash? I actually got a view into a similar state locally by hitting #9309 while working on views such that the view descriptor was persisted with the DB-qualified name in its name field. After attempting to drop it, the view is still in that awkward state where it doesn't show up anywhere but in `show tables`.

**Comment by vivekmenezes** [comment]
the table should get garbage collected after a crash, although I don't see a test testing that!

**Comment by nvanbenschoten** [comment]
This may be related to https://github.com/cockroachdb/cockroach/issues/8497.

**Comment by andreimatei** [comment]
@a-robinson, the process through which we're supposed to "garbage collect tables" is the async execution path for schema changes. Mutations on table descriptors are executed through either a sync path (blocking the DDL statement/transaction), and, if that fails, all the nodes are gonna race in getting a "schema application lease" and retrying executing it.
There's a test for this async path. @vivekmenezes did you have some other type of test in mind?
https://github.com/cockroachdb/cockroach/blob/da818b302edaaf3cb6d099178d14a02816ad8970/sql/schema_changer_test.go#L317

@nvanbenschoten we should fix #8497, and maybe it's related, but I think this present bug can also be explained as expected behavior (or at least I don't see why it's not expected).

**Comment by vivekmenezes** [comment]
the particular test you mention doesn't include a test for the Drop table case. We need to add that as part of resolving this issue.
